lib/Debugger.Protocol/String16.cpp

```diff
@@ -121,6 +127,69 @@ namespace JsDebug
         return String16(m_impl.substr(pos, len));
     }
 
+    std::string String16::toUtf8() const
+    {
+        static_assert(sizeof(wchar_t) == sizeof(UChar));
+
+        int requiredLength = WideCharToMultiByte(
+            CP_UTF8, 
+            0, 
+            reinterpret_cast<const wchar_t*>(m_impl.data()), 
+            static_cast<int>(m_impl.size()), 
+            nullptr, 
+            0, 
+            nullptr, 
+            nullptr);
+
+        if (requiredLength <= 0)
+        {
+            return std::string();
+        }
+
+        std::vector<char> utf8Chars(requiredLength);
+
+        if (WideCharToMultiByte(
+                CP_UTF8, 
+                0, 
+                reinterpret_cast<const wchar_t*>(m_impl.data()), 
+                static_cast<int>(m_impl.size()), 
+                utf8Chars.data(), 
+                static_cast<int>(utf8Chars.size()), 
+                nullptr, 
+                nullptr) == 0)
+        {
+            return std::string();
```

Comment:
I'm wondering whether returning an empty string is the right thing for this failure case. Of course, it doesn't seem like this should fail unless there is a problem with the input. According to the documentation these are the possible errors:

ERROR_INSUFFICIENT_BUFFER. A supplied buffer size was not large enough, or it was incorrectly set to NULL.
ERROR_INVALID_FLAGS. The values supplied for flags were not valid.
ERROR_INVALID_PARAMETER. Any of the parameter values was invalid. 
ERROR_NO_UNICODE_TRANSLATION. Invalid Unicode was found in a string.


The first three should not apply. If there's an invalid Unicode char, would it be better to skip the bad char(s) and only return the good ones? It could be only one char out of the whole string. But doing this may depend on the context. 

I'm not sure but I wanted to call this out.

Reply:
I don't think I can easily find out which character was bad without adding a bunch of logic.  If the string contains invalid characters then I don't think there should be any attempt to recover from it.  I toyed with the idea of exceptions as well, but it doesn't seem like there's a great place to handle them.

Reply:
Well, it would really help to know what is considered "Invalid Unicode" as I believe that even unassigned Unicode code points would be translatable to UTF-8 (as it isn't changing the character, just the byte representation of the code point). Even chars like U+FFFE and U+FFFF are translatable to UTF-8. It makes me wonder whether that error would occur at all, and if so, for which code points.

Is there any way for us to know whether this generates an error at all? I think what we are saying is that we don't believe it should really happen.

Reply:
It looks like the default behavior is to replace illegal sequences with `U+FFFD` unless the flag `*_ERR_INVALID_CHARS` is passed.  I think it's best to just keep using the default behavior.  Given that I'm not actually sure that conversion can fail unless the parameters are incorrect.

Reply:
Agreed. Thanks for looking into it.